### PR TITLE
Fix Issue 12357 - Deprecate calling text, wtext and dtext with 0 arguments

### PR DESCRIPTION
**Pull request by MetaLang**

### CODE REVIEW

**Comment by deadalnix** [comment]
This adds special casing. As discussed before, I'm against. It essentially pushes the checks on users, and add a special case.

**Comment by schveiguy** [comment]
While we're doing this, I thought of a more optimized implementation:

```D
         S result = "";
         foreach (arg; args)
             if(result.length) result ~= to!S(arg);
             else result = to!S(arg);
         return result;
```

This allows one to get equivalent performance to the zero arg version with `text("", args)`

**Comment by schveiguy** [comment]
Link to discussion: https://forum.dlang.org/post/skqcudmkvqtejmofxoim@forum.dlang.org

**Comment by JackStouffer** [comment]
@schveiguy I have a faster version here: https://github.com/dlang/phobos/pull/4280

**Comment by schveiguy** [comment]
@JackStouffer Thanks, I'll comment there. We can probably leave optimization of text out of this PR.

**Comment by schveiguy** [comment]
LGTM.

**Comment by MetaLang** [comment]
@andralex as this is technically a breaking change.

**Comment by ZombineDev** [comment]
I don't think this the right course of action. Just because a phobos band aid fix is easier doesn't mean it's the right thing to do. We don't need Scott Meyers to tell us on the next DConf that we should remove all uses of functions that may be callable with zero arguments from our code just because they may shadow  variables! This just insane! Just because this tiny change fixes a bug in your, doesn't mean that there aren't hundreds of other cases waiting to be discovered.

We need to fix the compiler instead.

**Comment by ZombineDev** [comment]
[deleted] sorry for noise, I misread the test case in the forum.

**Comment by JackStouffer** [comment]
I'm against this change, as it no longer allows a valid use case such as this:

```
auto someFunction(Args...)(Args args)
{
    ...
    enforce(someCondition, text(args));
}
```

You would now have to special case it

```
auto someFunction(Args...)(Args args)
{
    ...
    static if (Args.length > 0)
        enforce(someCondition, text(args));
    else
        enforce(someCondition, text("Some error"));
}
```

This is what I meant when I said `text` and friends are supposed to act as string constructors; they are designed to work with generic code.

**Comment by schveiguy** [comment]
```D
enforce(someCondition, text("", args));
```

**Comment by schveiguy** [comment]
FYI, @JackStouffer if you start your quote blocks with ```D, then github will syntax highlight for you.

**Comment by deadalnix** [comment]
Ok pretty much the same thing came up in the design of checked int, so I'll just copy my comment here. An empty sequence is a valid sequence and any function that accept a sequence you be able to handle it gracefully, even in its degenerate form.

Just like multiplying by 0 or 1 is not very useful, yet the MUL instruction and * operator accept these as operands. Like adding 0 is not useful, yet ADD and the + operator accept these as operand.

It is important for library code to handle degenerate cases gracefully.

@JackStouffer 's use case is legitimate, and I don't think this is the place of library author, especially standard lib to tell users they are wrong. Once you are in this business you've already lost.

**Comment by MetaLang** [comment]
On the other hand, how many `text = "asdf"` bugs happened before the bug was fixed with scoped imports shadowing local variables, function arguments and instance variables? A variant of this bug can still happen; I've shown it in the newsgroup thread. 

In my opinion you're putting far too much emphasis on an ideal abstract "elegance" in the language, but here we have a real bug that is hard to track down, and after this fix there will be a deprecation message and then a compiler error pointing to exactly where the code has gone wrong. To me the benefits outweigh the cons.

**Comment by schveiguy** [comment]
There are two ways to trigger this deprecation: `text` and `text(args)` where `args` is a tuple of 0 size. I posit that the former happens by accident quite frequently, `text` is a good name for a variable, so if you think your text variable is named that, but it's really named `Text` or `input` or something else, your code compiles and does strange things. Note that the former is always an error. The latter NEVER happens. Like literally never. What use case is "just convert all the sequence to text" and nothing else, even if it's empty? No leading message explaining what the rest of it means? The deprecation triggered zero errors in testing Phobos. I'm guessing that unit testing all of code.dlang.org would trigger zero errors as well.

Sorry guys, the use case simply doesn't exist, except in toy demonstrations.

**Comment by MetaLang** [comment]
Andrei has given his blessing: http://forum.dlang.org/post/nkn8kk$kes$1@digitalmars.com. Good to merge?

**Comment by deadalnix** [comment]
If we are going to proceed (which I still opposes), there need to be some kind of policy to know which function support an empty sequence and which do not. Randomly accepting this or not in various function is clowny.

@MetaLang No I do not. You don't get it. Yes this specific instance won't probably be that bad. But this one, plus that one plus the other one is really bad. And things are really bad right now. Every time there is a problem, we come up with a quick patch that solve that specific case by making the while edifice even more nonsensical. As a result @safe is not safe at all, immutable can be mutated (safely) TLS can be shared (still safely), enum can have value unexpected by final switch (safely once again) and it turtles all the way down. Special cases call for more special cases. It nevers ends. It is on and on and on. The very thing that started this discussion is special cases added previously to begin with. Seriously, when specials cases are the problem, how enlightened do you need to be to understand than more special cases is probably not the best way to takle this ?

I mean seriously, there are many way to handle this. More principled optional () is an option (unlikely to happen, but probably the right solution). Using a linter is an option. unittests is an option (really, that's all that is needed is the inital code presented). And I'm sure some people more imaginative than me have things to propose.

But no, we need to special case something in the standard library. That must be the sensible way forward.

I mean seriously, be consistent. Either special cases aren't a problem and the only things that this bug report deserve is RTFM, or specials cases are a problem, in which case it would probably be smart to not add more.

File: std/conv.d [hunk 1]
```diff
@@ -3675,11 +3675,25 @@ Target parseElement(Target, Source)(ref Source s)
  * Convenience functions for converting any number and types of
```
Comment (schveiguy) [inline std/conv.d]: Can you change "any number and types of arguments" to "one or more arguments of any type"?

**Comment by schveiguy** [comment]
> I mean seriously, be consistent. Either special cases aren't a problem and the only things that this bug report deserve is RTFM, or specials cases are a problem, in which case it would probably be smart to not add more.

What we have is a corner case that is never used, but often accidentally triggered by an error. We eliminate the pitfall, and everyone's code gets better.

Think of the "special case" of `if(condition);` Eliminating that possibility is surely not necessary, RTFM, right? But I'm grateful it's there, it's saved so many bugs, and the possibility to make the error added no value to the language.

**Comment by MetaLang** [comment]
Will rebase in a bit.

**Comment by MetaLang** [comment]
Rebased and ready to merge.

**Comment by schveiguy** [comment]
Auto-merge toggled on

**Comment by schveiguy** [comment]
Auto-merge toggled off

**Comment by schveiguy** [comment]
@MetaLang one minor thing -- there is a documented unit test that for some reason isn't getting into the docs (wasn't there before either). Can you fix? I think it's because the unittest is put below the impl version.

**Comment by schveiguy** [comment]
Auto-merge toggled on

**Comment by schveiguy** [comment]
Nevermind, I'll just fix it.

**Pull request merged**